File: FilterBankDesign.m
function [fb, cost, recon_err, iter] = FilterBankDesign(fb, eta, lambda, max_iter)
% Filter bank design tool
% Written by Xi-Lin Li, lixilinx@gmail.com
%
% INPUTS:
%   fb: a filter bank structure; empty fields will be assigned with default values assuming DFT modulation
%   eta: penalty coefficient on reconstruction errors
%   lambda: regularization factor on filter coefficient energy; set lambda>0 when Hessian is ill-conditioned
%   max_iter: maximum number of iterations
%
% OUTPUTS:
%   fb: a filter bank structure with designed values
%   cost: total design cost (stop band energies + reconstruction errors + errors due to violation of constraints)
%   recon_err: average reconstruction error
%   iter: number of iterations performed
%
if isempty(fb.T) || isempty(fb.B) || isempty(fb.tau0) || isempty(fb.h) || isempty(fb.g)
    error('Period of modulation sequences, decimation ratio, system delay, and initial guesses for prototype filters must be provided\n');
end
if isempty(fb.Gamma)
    fb.Gamma = speye(fb.T);
end
if isempty(fb.i)
    fb.i = mod(-fb.tau0, fb.T);
end
if isempty(fb.j)
    fb.j = 0;
end
if isempty(fb.w_cut)
    fb.w_cut = pi/fb.B;
end
if isempty(fb.zeta)
    fb.zeta = 1;
end
if isempty(fb.symmetry)
    fb.symmetry = [0;0;0];
end
if isempty(fb.momentum)
    fb.momentum = 0;
end
   
Gamma=fb.Gamma; T=fb.T; B=fb.B; tau0=fb.tau0; shift_i=fb.i; shift_j=fb.j; h=fb.h(:); g=fb.g(:); w_cut=fb.w_cut; zeta=fb.zeta; symmetry=fb.symmetry(:); momentum=fb.momentum;
Lh = length(h); Lg = length(g);

if T<0 || B<0 || w_cut<0 || w_cut>=pi || zeta<0 || eta<=0 || lambda<0 || length(symmetry)~=3 || B>T || tau0<0 || tau0>Lh+Lg-2 || (symmetry(1)~=0 && Lh~=Lg) || Lh<B || Lg<B
    error('Invalid filter bank design settings\n');
end

% check (fb.i, fb.j)'s feasibility
valid_t_tau = get_valid_t_tau(Gamma, T, B, Lh, Lg, shift_i, shift_j);
if size(valid_t_tau, 1) < B
    error('Pair (fb.i, fb.j) is infeasible');
else if sum(valid_t_tau(:,2)==tau0) < B
        error('Pair (fb.i, fb.j) is infeasible');
    end
end

Tol = 1e-5; % tolerance condition for convergence
for iter = 1 : max_iter
    [cost, grad_h, grad_g, recon_err] = fbd_cost_grad(Gamma, shift_i, shift_j, valid_t_tau, T, B, tau0, h, g, w_cut, zeta, symmetry, momentum, eta, lambda, 1, (iter==1));
    
    new_cost = inf;
    step = 1/2;
    while new_cost >= cost && step >= 1/2^16 % if steps 1/2^2, 1/2^4, 1/2^8 all fails, we accept 1/2^16 anyway
        new_h = h - step*grad_h;
        new_g = g - step*grad_g;
        [new_cost, ~, ~, ~] = fbd_cost_grad(Gamma, shift_i, shift_j, valid_t_tau, T, B, tau0, new_h, new_g, w_cut, zeta, symmetry, momentum, eta, lambda, 0, 0);
        step = step*step;
    end
    h = new_h;
    g = new_g;
    if max(max(abs(grad_h)), max(abs(grad_g))) < Tol
        break;
    end
end
fb.Gamma=Gamma; fb.T=T; fb.B=B; fb.tau0=tau0; fb.i=shift_i; fb.j=shift_j; fb.h=h; fb.g=g; fb.w_cut=w_cut; fb.zeta=zeta; fb.symmetry=symmetry;


function [cost, grad_h, grad_g, recon_err] = fbd_cost_grad(Gamma, shift_i, shift_j, valid_t_tau, T, B, tau0, h, g, w_cut, zeta, symmetry, momentum, eta, lambda, need_grad, init)
% cost and gradients evaluation for filter bank design
%
% INPUTS:
%  	filter bank structure fields, and penalty factors eta and lambda
%  	need_grad: 0 if no need to evaluate gradients; 1 if need gradients
%   init:   1 for initialization; 0 for no initialization
%
% OUTPUTS:
%  	cost: cost
% 	-grad_h: search direction for analysis filter coefficients
%  	-grad_g: search direction for synthesis filter coefficients
% 	recon_err: average reconstruction error
%
persistent hess_fixed_part; % fixed part of Hessian
persistent all_Mask; % the M(t, tau) matrices, and one more matrix for constraint h'*h=g'*g
persistent all_delta_tau_tau0; % all delta(tau - tau0), and one more zero for constraint h'*h=g'*g

x = [h; g]; % x is the theta in report
Lh = length(h);
Lg = length(g);

if init || isempty(hess_fixed_part) || isempty(all_Mask) || isempty(all_delta_tau_tau0)
    hess_fixed_part = blkdiag(matrix_stopband_energy(Lh, w_cut, momentum), zeta*matrix_stopband_energy(Lg, w_cut, momentum));
    hess_fixed_part = hess_fixed_part + lambda*eye(Lh + Lg);
    if symmetry(1)>0
        hess_fixed_part = hess_fixed_part + eta*[eye(Lh), -fliplr(eye(Lh)); -fliplr(eye(Lh)), eye(Lh)];
    else if symmetry(1)<0
            hess_fixed_part = hess_fixed_part + eta*[eye(Lh), -eye(Lh); -eye(Lh), eye(Lh)];
        end
    end
    if symmetry(2)
        hess_fixed_part(1:Lh, 1:Lh) = hess_fixed_part(1:Lh, 1:Lh) + eta*(eye(Lh) - fliplr(eye(Lh)));
    end
    if symmetry(3)
        hess_fixed_part(Lh+1:Lh+Lg, Lh+1:Lh+Lg) = hess_fixed_part(Lh+1:Lh+Lg, Lh+1:Lh+Lg) + eta*(eye(Lg) - fliplr(eye(Lg)));
    end
    
    Mask_cnt = 0;
    all_Mask = cell(size(valid_t_tau, 1) + 1, 1); % the last one is for constraint h'*h=g'*g
    all_delta_tau_tau0 = zeros(1, size(valid_t_tau, 1) + 1); % the last zero is for h'*h=g'*g; we just put it in delta(tau-tau0)
    for t_tau = 1 : size(valid_t_tau, 1)
        t = valid_t_tau(t_tau, 1);
        tau = valid_t_tau(t_tau, 2);
        
        n0 = max(ceil((t-tau)/B), floor((t-Lg)/B)+1);
        n1 = min(floor(t/B), ceil((Lh+t-tau)/B)-1);
        i = zeros(2*(n1-n0+1), 1);
        j = zeros(2*(n1-n0+1), 1);
        s = zeros(2*(n1-n0+1), 1);
        k = 0;
        for n = n0 : n1
            i(k+1) = n*B+tau-t +1;
            j(k+1) = t-n*B + Lh + 1;
            s(k+1) = 0.5*Gamma(mod(t-n*B+shift_j, T) + 1, mod(t-tau-n*B-shift_i, T) + 1);
            i(k+2) = j(k+1);
            j(k+2) = i(k+1);
            s(k+2) = s(k+1);
            k = k + 2;
        end
        Mask_cnt = Mask_cnt + 1;
        all_Mask{Mask_cnt} = sparse(i, j, s, Lh+Lg, Lh+Lg);
        
        if tau == tau0
            all_delta_tau_tau0(Mask_cnt) = 1;
        end
    end
    
    all_Mask{end} = blkdiag(eye(Lh), -eye(Lg)); % this last one is for h'*h-g'*g
end

all_Mx = zeros(Lh+Lg, length(all_Mask));
for i = 1 : length(all_Mask)
    all_Mx(:, i) = all_Mask{i}*x;
end

errs = x'*all_Mx - all_delta_tau_tau0;
recon_err = errs(1:end-1)*errs(1:end-1)'/B; % the last one is h'*h - g'*g
cost = 0.5*x'*hess_fixed_part*x + 0.5*eta*(errs*errs');

if need_grad
    grad = hess_fixed_part*x + 2*eta*all_Mx*errs';
    grad = (hess_fixed_part + 4*eta*(all_Mx*all_Mx'))\grad;
else
    grad = zeros(size(x));
end
grad_h = grad(1:Lh);
grad_g = grad(Lh+1:end);


function P = matrix_stopband_energy(L, w_cut, momentum)
% the Toeplitz matrix \Pi in the report
%   L: size of P
%   w_cut: cutoff angular frequency
%   momentum: 0, 1, or 2
%
h = zeros(1, L);
pmq = 1 - (2:L); % p - q

switch momentum
    case 0
        h(1) = pi - w_cut;
        h(2:end) = -sin( pmq*w_cut )./pmq;
    case 1
        h(1) = (pi^2 - w_cut^2)/2;
        h(2:end) = -( cos(pmq*w_cut) + w_cut*pmq.*sin(pmq*w_cut) - (-1).^pmq )./pmq.^2;
    case 2
        h(1) = (pi^3 - w_cut^3)/3;
        h(2:end) = -( (w_cut^2*pmq.^2 - 2).*sin(pmq*w_cut) + 2*w_cut*pmq.*cos(pmq*w_cut) - 2*pi*pmq.*(-1).^pmq )./pmq.^3;
    otherwise
        error('Design momentum undefined. You may specify your own design here.');
end

P = toeplitz( h );


function valid_t_tau = get_valid_t_tau(Gamma, T, B, Lh, Lg, shift_i, shift_j)
% to get the list of valid t's and tau's, i.e., (t, tau) such that e(t, tau)~=0
valid_t_tau = [];
for t = 0 : B-1
    for tau = 0 : Lh+Lg-2
        n0 = max(ceil((t-tau)/B), floor((t-Lg)/B)+1);
        n1 = min(floor(t/B), ceil((Lh+t-tau)/B)-1);
        for n = n0 : n1
            i = mod(t-n*B+shift_j, T) + 1;
            j = mod(t-tau-n*B-shift_i, T) + 1;
            if Gamma(i,j) ~= 0
                valid_t_tau = [valid_t_tau; [t, tau]];
                break;
            end
        end
    end
end

File: filterbank_design.py
"""Periodic-sequence-modulated filterbank design. 
Rewritten with AI based on materials:
    1, https://ieeexplore.ieee.org/document/8304771.
    2, https://github.com/lixilinx/FilterbanksBestPractices/blob/main/FilterBankDesign.m. 
"""

from dataclasses import dataclass, replace

import numpy as np
from numpy.typing import ArrayLike, NDArray
from scipy import linalg, sparse

FloatArray = NDArray[np.float64]
SparseMatrix = sparse.coo_matrix

@dataclass
class FilterBank:
    """The filterbank object. 
    """
    T: int # modulation sequence period
    B: int # decimation ratio/block size/hop size
    tau0: int # system delay
    h: ArrayLike # analysis prototype filter
    g: ArrayLike # synthesis prototype filter
    Gamma: ArrayLike | sparse.spmatrix | None = None # modulation product matrix
    i: int | None = None # analysis modulation shift
    j: int | None = None # synthesis modulation shift
    w_cut: float | None = None # prototype cutoff in rad/sample 
    zeta: float | None = None # synthesis stop-band weight
    symmetry: ArrayLike | None = None # symmetry[0]=1: g=flip(h)/MIRROR; symmetry[0]=-1: g=h/SAME; symmetry[1]=1: h=flip(h); symmetry[2]=1: g=flip(g)
    momentum: int | None = None # stop-band frequency weighting: 0, 1, or 2

    def __post_init__(self) -> None:
        assert np.ndim(self.h) == 1 and np.ndim(self.g) == 1
        Lh, Lg = np.size(self.h), np.size(self.g)
        assert self.T > 0
        assert 0 < self.B <= self.T
        assert Lh >= self.B and Lg >= self.B, "h/g must have >= B taps"
        assert self.B - 1 <= self.tau0 <= Lh + Lg - self.B - 1, "infeasible delay" # tighter bounds than matlab code 
        if self.Gamma is not None:
            assert self.Gamma.shape == (self.T, self.T)
        if self.w_cut is not None:
            assert 0 <= self.w_cut < np.pi 
        if self.zeta is not None:
            assert self.zeta >= 0
        if self.symmetry is not None:
            symmetry = np.asarray(self.symmetry).reshape(-1)
            assert symmetry.size == 3
            assert symmetry[0] == 0 or Lh == Lg, "MIRROR/SAME symmetry requires equal filter length"
        if self.momentum is not None:
            assert self.momentum in (0, 1, 2), "only momentum 0, 1, and 2 implemented"


@dataclass(frozen=True)
class DesignResult:
    """Hold design result returned by func design_filter_bank. 
    """
    filter_bank: FilterBank
    cost: float
    reconstruction_error: float
    iterations: int
    converged: bool


def _normalize_filter_bank(filter_bank: FilterBank) -> FilterBank:
    """Fill defaults. 
    """
    T, B, tau0 = filter_bank.T, filter_bank.B, filter_bank.tau0
    h = np.asarray(filter_bank.h, dtype=float).reshape(-1).copy()
    g = np.asarray(filter_bank.g, dtype=float).reshape(-1).copy()
    Gamma = filter_bank.Gamma
    if Gamma is None:
        Gamma = sparse.eye(T, format="csr")
    elif sparse.issparse(Gamma):
        Gamma = Gamma.tocsr().astype(float, copy=True)
    else:
        Gamma = np.asarray(Gamma, dtype=float).copy()

    return replace(
        filter_bank,
        Gamma=Gamma,
        i=(-tau0) % T if filter_bank.i is None else filter_bank.i,
        j=0 if filter_bank.j is None else filter_bank.j,
        h=h, g=g,
        w_cut=np.pi/B if filter_bank.w_cut is None else filter_bank.w_cut, # w_cut=pi if B=1, bad! 
        zeta=1.0 if filter_bank.zeta is None else filter_bank.zeta,
        symmetry=np.zeros(3) if filter_bank.symmetry is None else np.asarray(filter_bank.symmetry, dtype=float).reshape(-1).copy(),
        momentum=0 if filter_bank.momentum is None else filter_bank.momentum,
    )


def _ceil_div(numerator: int, denominator: int) -> int:
    return -((-numerator) // denominator)


def _active_terms(t: int, tau: int, B: int, Lh: int, Lg: int) -> range:
    """Return frame indices whose filter taps contribute at (t, tau).
    """
    n0 = max(_ceil_div(t - tau, B), (t - Lg) // B + 1)
    n1 = min(t // B, _ceil_div(Lh + t  - tau, B) - 1)
    return range(n0, n1 + 1)


def _valid_t_tau(Gamma: FloatArray, T: int, B: int, Lh: int, Lg: int, shift_i: int, shift_j: int) -> list[tuple[int, int]]:
    """Return all the (t, tau) pairs where M(t, tau) is not a zero matrix. 
    The Matlab code is greatly simplify after re-ordering the loop indices. 
    """
    tau_count = Lh + Lg - 1
    gamma_cols = (-np.arange(Lh) - shift_i) % T
    row_hits = Gamma[:, gamma_cols] != 0.0
    valid: list[tuple[int, int]] = []
    for t in range(B):
        hit = np.zeros(tau_count, dtype=bool)
        for n in range((t - Lg) // B + 1, t // B + 1):
            first = t - n * B
            hit[first : first + Lh] |= row_hits[(first + shift_j) % T]
        valid.extend((t, int(tau)) for tau in np.flatnonzero(hit))
    return valid


def _constraint_masks(Gamma: FloatArray, T: int, B: int, Lh: int, Lg: int, shift_i: int, shift_j: int, 
                      valid_t_tau: list[tuple[int, int]], tau0: int) -> tuple[SparseMatrix, FloatArray]:
    """Return all the M/mask matrices and targets. Mask matrices are row-stacked. 
    """
    size = Lh + Lg
    mask_count = len(valid_t_tau) + 1 # extra mask matrix for term h'h - g'g 
    targets = np.zeros(mask_count, dtype=np.float64)
    rows: list[int] = []
    cols: list[int] = []
    values: list[float] = []

    for mask_index, (t, tau) in enumerate(valid_t_tau):
        offset = mask_index * size
        for n in _active_terms(t, tau, B, Lh, Lg):
            h_index = n * B + tau - t
            g_index = Lh + t - n * B
            gamma_row = (t - n * B + shift_j) % T
            gamma_col = (t - tau - n * B - shift_i) % T
            value = 0.5 * Gamma[gamma_row, gamma_col]
            if value != 0.0:
                rows.extend((offset + h_index, offset + g_index))
                cols.extend((g_index, h_index))
                values.extend((value, value))
        if tau == tau0:
            targets[mask_index] = 1.0

    # the last mask blkdiag(I, -I) for the energy-balance term 
    offset = len(valid_t_tau) * size
    rows.extend(range(offset, offset + size))
    cols.extend(range(size))
    values.extend([1.0] * Lh + [-1.0] * Lg)

    stacked = sparse.coo_matrix((values, (rows, cols)), shape=(mask_count * size, size), dtype=np.float64)
    return stacked, targets


def _stopband_energy_matrix(length: int, w_cut: float, momentum: int) -> FloatArray:
    """Return the Toeplitz stop-band energy matrix. 
    """
    first_row = np.empty(length, dtype=np.float64)
    lag = 1.0 - np.arange(2, length + 1, dtype=np.float64)
    if momentum == 0:
        first_row[0] = np.pi - w_cut
        first_row[1:] = -np.sin(lag * w_cut) / lag
    elif momentum == 1:
        first_row[0] = (np.pi**2 - w_cut**2) / 2.0
        first_row[1:] = -(np.cos(lag * w_cut) + w_cut * lag * np.sin(lag * w_cut) - np.power(-1.0, lag)) / lag**2
    else: # momentum=2
        first_row[0] = (np.pi**3 - w_cut**3) / 3.0
        first_row[1:] = -((w_cut**2 * lag**2 - 2.0) * np.sin(lag * w_cut) + 2.0 * w_cut * lag * np.cos(lag * w_cut) - 2.0 * np.pi * lag * np.power(-1.0, lag)) / lag**3
    return linalg.toeplitz(first_row)


def _fixed_hessian(filter_bank: FilterBank, eta: float, lambda_: float) -> FloatArray:
    """Return the fixed part of the Hessian. 
    """
    h, g = np.asarray(filter_bank.h), np.asarray(filter_bank.g)
    symmetry = np.asarray(filter_bank.symmetry)
    w_cut = float(filter_bank.w_cut)
    zeta = float(filter_bank.zeta)
    momentum = int(filter_bank.momentum)
    Lh, Lg = h.size, g.size
    size = Lh + Lg

    hessian = np.zeros((size, size), dtype=np.float64)
    hessian[:Lh, :Lh] = _stopband_energy_matrix(Lh, w_cut, momentum)
    hessian[Lh:, Lh:] = zeta * _stopband_energy_matrix(Lg, w_cut, momentum)

    h_tap, g_tap = np.arange(Lh), np.arange(Lg)
    h_mirror, g_mirror = Lh - 1 - h_tap, Lg - 1 - g_tap
    all_taps = np.arange(size)
    hessian[all_taps, all_taps] += lambda_

    if symmetry[0] != 0.0:
        partner = h_mirror if symmetry[0] > 0.0 else h_tap
        hessian[h_tap, h_tap] += eta
        hessian[Lh + h_tap, Lh + h_tap] += eta
        hessian[h_tap, Lh + partner] -= eta
        hessian[Lh + partner, h_tap] -= eta

    h_paired = h_tap[h_tap != h_mirror] # exclude central tap for odd length filter 
    g_paired = g_tap[g_tap != g_mirror]
    if symmetry[1] != 0.0:
        hessian[h_paired, h_paired] += eta
        hessian[h_paired, Lh - 1 - h_paired] -= eta
    if symmetry[2] != 0.0:
        hessian[Lh + g_paired, Lh + g_paired] += eta
        hessian[Lh + g_paired, Lh + Lg - 1 - g_paired] -= eta

    return hessian


def _evaluate(x: FloatArray, fixed_hessian: FloatArray, stacked_masks: SparseMatrix, targets: FloatArray, B: int, 
              eta: float) -> tuple[float, float, FloatArray, FloatArray]:
    """Evaluate a filterbank prototype filter design x = (h, g).
    """
    all_mx = (stacked_masks @ x).reshape(-1, x.size).T
    errors = x @ all_mx - targets
    reconstruction_error = float(errors[:-1] @ errors[:-1] / B) # exclude the last one (the energy-balance term) 
    cost = float(0.5 * x @ fixed_hessian @ x + 0.5 * eta * (errors @ errors))
    return cost, reconstruction_error, errors, all_mx


def _search_direction(x: FloatArray, fixed_hessian: FloatArray, errors: FloatArray, all_mx: FloatArray, eta: float) -> FloatArray:
    """Compute the Hessian-preconditioned search direction. 
    """
    gradient = fixed_hessian @ x + 2.0 * eta * (all_mx @ errors)
    system = all_mx @ all_mx.T
    system *= 4.0 * eta 
    system += fixed_hessian
    try:
        direction = linalg.solve(system, gradient, assume_a="pos", check_finite=False) # Cholesky 
    except linalg.LinAlgError as error:
        raise np.linalg.LinAlgError("Hessian not positive definite; increase lambda_") from error 
    return np.asarray(direction, dtype=np.float64)


def design_filter_bank(filter_bank: FilterBank, eta: float, 
                       lambda_: float = 0.0, max_iterations: int = 100, tolerance: float = 1e-5) -> DesignResult:
    """Design prototype filters. 
    eta: penalty coefficient for reconstruction and constraints. 
    lambda_: L2 regularization coefficient. 
    """
    fb = _normalize_filter_bank(filter_bank)
    h, g = np.asarray(fb.h), np.asarray(fb.g)
    Gamma = fb.Gamma
    T, B, tau0, shift_i, shift_j = map(int, (fb.T, fb.B, fb.tau0, fb.i, fb.j))
    Lh, Lg = h.size, g.size

    gamma_lookup = Gamma.toarray() if sparse.issparse(Gamma) else np.asarray(Gamma)

    valid_t_tau = _valid_t_tau(gamma_lookup, T, B, Lh, Lg, shift_i, shift_j)
    if sum(tau == tau0 for _, tau in valid_t_tau) < B:
        raise ValueError("the given (i, j)-shift pair is infeasible")

    stacked_masks, targets = _constraint_masks(gamma_lookup, T, B, Lh, Lg, shift_i, shift_j, valid_t_tau, tau0)
    fixed_hessian = _fixed_hessian(fb, eta, lambda_)
    x = np.concatenate((h, g))
    converged = False 
    iterations = 0
    step_candidates = (1/2, 1/4, 1/16, 1/256, 1/65536)

    for iteration in range(1, max_iterations + 1):
        cost, _, errors, all_mx = _evaluate(x, fixed_hessian, stacked_masks, targets, B, eta)
        direction = _search_direction(x, fixed_hessian, errors, all_mx, eta)

        candidate = x
        for step in step_candidates:
            candidate = x - step * direction
            candidate_cost, *_ = _evaluate(candidate, fixed_hessian, stacked_masks, targets, B, eta)
            if candidate_cost < cost:
                break

        x = candidate
        iterations = iteration
        if np.max(np.abs(direction)) < tolerance:
            converged = True
            break

    final_cost, final_reconstruction_error, *_ = _evaluate(x, fixed_hessian, stacked_masks, targets, B, eta)
    designed = replace(fb, h=x[:Lh].copy(), g=x[Lh:].copy())
    return DesignResult(filter_bank=designed, cost=final_cost, reconstruction_error=final_reconstruction_error,
                        iterations=iterations, converged=converged)
